protocols/dcutr/src/handler/relayed.rs

```diff
@@ -0,0 +1,380 @@
+                        protocol: SubstreamProtocol::new(
+                            protocol::outbound::Upgrade::new(obs_addrs),
+                            attempt,
+                        ),
+                    });
+            }
+            Command::AcceptInboundConnect {
+                inbound_connect,
+                obs_addrs,
+            } => {
+                self.inbound_connects
+                    .push(inbound_connect.accept(obs_addrs).boxed());
+            }
+            Command::UpgradeFinishedDontKeepAlive => {
+                self.keep_alive = KeepAlive::No;
+            }
+        }
+    }
+
+    fn inject_listen_upgrade_error(
+        &mut self,
+        _: Self::InboundOpenInfo,
+        error: ProtocolsHandlerUpgrErr<<Self::InboundProtocol as InboundUpgradeSend>::Error>,
+    ) {
+        self.keep_alive = KeepAlive::No;
+
+        match error {
+            ProtocolsHandlerUpgrErr::Timeout => {
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::InboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Timeout,
+                    },
+                ));
+            }
+            ProtocolsHandlerUpgrErr::Timer => {
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::InboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Timer,
+                    },
+                ));
+            }
+            ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)) => {
+                // The remote merely doesn't support the DCUtR protocol.
+                // This is no reason to close the connection, which may
+                // successfully communicate with other protocols already.
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::InboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                            NegotiationError::Failed,
+                        )),
+                    },
+                ));
+            }
+            _ => {
+                // Anything else is considered a fatal error or misbehaviour of
+                // the remote peer and results in closing the connection.
+                self.pending_error = Some(error.map_upgrade_err(|e| {
+                    e.map_err(|e| match e {
+                        EitherError::A(e) => EitherError::A(e),
+                        EitherError::B(v) => void::unreachable(v),
+                    })
+                }));
+            }
+        }
+    }
+
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _open_info: Self::OutboundOpenInfo,
+        error: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgradeSend>::Error>,
+    ) {
+        self.keep_alive = KeepAlive::No;
```

Comment:
Today the retry mechanism in DCUtR retries on hole-punch failure only. E.g. when the direct connection establishment itself fails.

`inject_dial_upgrade_error` called on `relayed::Handler` implies that the communication via the relayed connection failed, not the hole-punch.

I would argue that there is not much use in retrying a _direct connection upgrade_ over a erroneous relayed connection. Would you agree with that reasoning @MarcoPolo?

Reply:
is the connection terminally failed? As in it may not recover in the future? If so, then I think it's okay to have a KeepAlive::No;. Otherwise Wouldn't it make sense to only set it in the terminal error case? 

For example if this is just a timeout, maybe we want to try again.

Reply:
Just to avoid any confusion. setting `KeepAlive::No` signals that this handler has no more use for the connection. In case another handler returns `KeepAlive::Yes` the connection is still kept alive.

> is the connection terminally failed?

No. One could still reuse the connection.

> For example if this is just a timeout, maybe we want to try again.

We might want to do that in the future, but today there is no mechanism that will trigger a retry. In other words, DCUtR will never use the connection again. Given that it never uses the connection again, it should set `KeepAlive::No`.

Does the above make sense @MarcoPolo? Am I missing something?

Reply:
oh!! I realized my confusion this is in the relayed handler that is trying to orchestrate the dance before we start the simultaneous tcp connections. For some reason I thought this was part of the rety logic during the simultaneous tcp connections.

Thanks, and sorry for the confusion.

Reply:
No worries. Actually I had the same thought as well when I first read your initial comment.